lib/class-wp-rest-widget-types-controller.php

```diff
@@ -0,0 +1,456 @@
+	}
+
+	/**
+	 * Registers the routes for the objects of the controller.
+	 *
+	 * @since x.x.0
+	 *
+	 * @see register_rest_route()
+	 */
+	public function register_routes() {
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+					'args'                => $this->get_collection_params(),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<name>[a-zA-Z0-9_-]+)',
+			array(
+				'args' => array(
+					'name'     => array(
+						'description' => __( 'Name of the widget.', 'gutenberg' ),
+						'type'        => 'string',
+						'required'    => true,
+					),
+					'instance' => array(
+						'description' => __( 'Current widget instance', 'gutenberg' ),
+						'type'        => 'object',
+						'default'     => array(),
+					),
+				),
+				array(
+					'methods'             => WP_REST_Server::EDITABLE,
+					'permission_callback' => array( $this, 'get_item_permissions_check' ),
+					'callback'            => array( $this, 'get_widget_form' ),
+					'args'                => array(
+						'context' => $this->get_context_param( array( 'default' => 'edit' ) ),
+					),
+				),
+			)
+		);
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<name>[a-zA-Z0-9_-]+)',
+			array(
+				'args'   => array(
+					'name' => array(
+						'description' => __( 'Widget name.', 'gutenberg' ),
+						'type'        => 'string',
+					),
+				),
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_item' ),
+					'permission_callback' => array( $this, 'get_item_permissions_check' ),
+					'args'                => $this->get_collection_params(),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Checks whether a given request has permission to read post widget types.
+	 *
+	 * @since x.x.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_Error|bool True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		return $this->check_read_permission();
+	}
+
+	/**
+	 * Retrieves all post widget types, depending on user context.
+	 *
+	 * @since x.x.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		global $wp_widget_factory;
+
+		$data = array();
+		foreach ( $wp_widget_factory->widgets as $slug => $widget ) {
```

Comment:
This needs to support widgets registered with `wp_register_sidebar_widget()`. See `::do_test_update_item_legacy_widget`.

Reply:
I don't this it needs to. This is a endpoint for widget types. `wp_register_sidebar_widget` does not register a widget type, it registers a instance of a widget. Unless I create this code wrong.

Reply:
Then where does the `description` live? Will we duplicate that field in the widgets controller?

Reply:
It lives here. https://github.com/WordPress/gutenberg/pull/26042/files#diff-a6036d6ccb63c98c7f607f1bcc572b84a185f194e3668d747f37123ba1207e60R370

And yes, it will be repeated fields. Which is why I wanted to add this now, so we can remove those fields from the widget controller.

Reply:
That's not what I'm saying. Where will the `description` field for widgets registered with `wp_register_sidebar_widget` live?

If we remove the repeated fields from the widgets controller, then there will be no way to get legacy widget's descriptions since they won't have an entry in the types controller.

Reply:
Oh, yeah. I guess we would have to leave that in the widget controller.

Reply:
This would also effect all of the other arguments for `widget_options` as well as for controls registered with `wp_register_widget_control`. So I don't see how we can meaningfully implement a separate types controller without support for legacy widgets as long as we are going to support legacy widgets at all.

Reply:
I think we should be able to iterate over the ones registered with `wp_register_sidebar_widget` and exclude ones that have an id base that correspond to an entry in the widgets factory.

Reply:
I don't see widget registered with `wp_register_sidebar_widget` as a valid widget type. It is setup up a instance of a widget, not a widget type. There are two different things. I don't think they should be support here, as it does make any sense and the fields do not match at all. WP_Widget has been around for a LONG time, since 2.8.0. I am also not overly worried about `wp_register_sidebar_widget`. Anyone using a widget like that has likely not upgraded.

Reply:
Gutenberg currently supports it both in the UI and the server. Dropping support for that would have to be done in a separate issue.

I'm also not sure how the fields are incompatible, they should be identical. `WP_Widget::_register_one` registers itself as a widget with the defined widget options in the constructor.

And the `control_options` in the `WP_Widget` constructor are defined by referencing the ones supported by `wp_register_widget_control`.

Reply:
Fixed in https://github.com/WordPress/gutenberg/pull/26042/commits/f18fc367184fbb98b25089034528b487cc88f2ad

Reply:
This is great, thank you! Could we add a test copying the widget definition from `do_test_update_item_legacy_widget`?

Reply:
Sorry, totally missed you added that already.